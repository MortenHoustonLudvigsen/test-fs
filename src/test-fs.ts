import './methods';

export { Entry, Stats, isEntry } from './Entry';
export * from './TestFs';
