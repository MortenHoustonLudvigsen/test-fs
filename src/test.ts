import { TestFs, Entry, isEntry } from 'test-fs';
import { resolve as resolvePath, relative as relativePath } from 'path';
import { inspect } from 'util';

const pkgDir: (cwd: string) => string = require('pkg-dir').sync;
export const rootDir = pkgDir(resolvePath(__dirname));
export const testFilesDir = resolvePath(rootDir, 'test-files');
export const testPaths = {
    physicalDir: testFilesDir,
    physicalFile: resolvePath(testFilesDir, 'file.txt'),
    physicalFile2: resolvePath(testFilesDir, 'file2.txt'),
    physicalBinaryFile: resolvePath(testFilesDir, 'file.bin'),
    emptyPhysicalDir: resolvePath(testFilesDir, 'emptydir'),
    virtualFileInPhysicalFile: resolvePath(testFilesDir, 'file.txt', 'dummy.txt'),
    virtualDir: resolvePath(testFilesDir, 'dummyDir'),
    virtualFile: resolvePath(testFilesDir, 'dummy.txt'),
    virtualFileInVirtualDir: resolvePath(testFilesDir, 'dummyDir/dummy.txt'),
    virtualDirInVirtualDir: resolvePath(testFilesDir, 'dummyDir/dummyDir2'),
}

function formatPath(path: string) {
    path = relativePath(rootDir, path).replace(/\\+/g, '/');
    if (path === '') {
        path = '.';
    }
    return path;
}

function serializeEntry(entry: Entry, indent: string, first: boolean): string[] {
    function getValue(key: string) {
        switch (key) {
            case 'path':
                return formatPath(entry.path);
            case 'dir':
                return typeof entry.dir === 'string' ? inspect(formatPath(entry.dir)) : undefined;
            case 'mode':
                return Number.isInteger(entry.mode!) ? `0o${entry.mode!.toString(8)}` : undefined;
            default:
                const value = (<any>entry)[key];
                return typeof value !== 'undefined' ? inspect(value) : value;
        }
    }

    function add(...keys: string[]) {
        for (let key of keys) {
            const value = getValue(key);
            if (typeof value === 'undefined') {
                return;
            }
            result.push(`${indent}    ${pad(key, 9, '.')}: ${value}`);
        }
    }

    const result: string[] = [];

    if (!first) {
        result.push('');
    }
    result.push(`${indent}<${getValue('path')}>`);
    add('name', 'dir', 'type', 'version', 'mode', 'statCount', 'readCount', 'size', 'virtual');

    switch (entry.type) {
        case 'initial':
            break;
        case 'directory':
            Object.keys(entry.entries).sort().forEach(name => {
                result.push(...serializeEntry(entry.entries[name], indent + '    ', false))
            });
            break;
        case 'file':
            add('contents');
            break;
        case 'none':
            break;
    }

    return result;
}

expect.addSnapshotSerializer({
    test: value => isEntry(value),
    print: (entry: Entry) => {
        return serializeEntry(entry, '', true).join('\n');
    }
});

expect.addSnapshotSerializer({
    test: value => Array.isArray(value) && value.every(v => typeof v === 'number'),
    print: (value: number[]) => {
        return JSON.stringify(value);
    }
});

const debugAllEvents = false;
const debugNoEvents = true;

function pad(str: string, length: number, char = ' ') {
    while (str.length < length) {
        str = str + char;
    }
    return str;
}

function entryType(entry: Entry) {
    switch (entry.type) {
        case 'initial':
            return 'initial';
        case 'directory':
            return entry.virtual ? 'virtual dir' : 'physical dir';
        case 'file':
            return entry.virtual ? 'virtual file' : 'physical file';
        case 'none':
            return entry.virtual ? 'virtual none' : 'physical none';
    }
}

export function formatStats(entry: Entry) {
    return `[v${entry.version} - ${entry.statCount} stats - ${entry.readCount} reads] ${pad(entryType(entry), 13)} ${entry.path}`;
}

export function createFs(debugEvents = debugAllEvents, indent: number = 0, ...events: string[]) {
    const fs = new TestFs();
    events = events.filter(e => !!e);
    if (!debugNoEvents && debugEvents) {
        fs.on('any', (event, entry) => {
            if (debugAllEvents || events.length === 0 || events.indexOf(event) >= 0) {
                console.log(`${pad('', indent)}Event: ${pad(event, 12)} ${formatStats(entry)}`);
            }
        });
    }
    return fs;
}

export type SimpleCallback<TResult> = (result?: TResult) => any;
export function simplePromise<TResult>(self: any, fn: (callback: SimpleCallback<TResult>) => any): Promise<[any, TResult | undefined]>;
export function simplePromise<TArg1, TResult>(self: any, fn: (arg1: TArg1, callback: SimpleCallback<TResult>) => any, arg1: TArg1): Promise<[any, TResult | undefined]>;
export function simplePromise<TArg1, TArg2, TResult>(self: any, fn: (arg1: TArg1, arg2: TArg2, callback: SimpleCallback<TResult>) => any, arg1: TArg1, arg2: TArg2): Promise<[any, TResult | undefined]>;
export function simplePromise<TResult>(self: any, fn: (...args: any[]) => any, ...args: any[]): Promise<[any, TResult | undefined]> {
    return new Promise((resolve, reject) => {
        fn.call(self, ...args, (result: TResult) => {
            resolve([undefined, result]);
        });
    });
}

export type Callback<TResult> = (err?: any, result?: TResult) => any;
export function promise<TResult>(self: any, fn: (callback: Callback<TResult>) => void | any): Promise<[any, TResult | undefined]>;
export function promise<TArg1, TResult>(self: any, fn: (arg1: TArg1, callback: Callback<TResult>) => void | any, arg1: TArg1): Promise<[any, TResult | undefined]>;
export function promise<TArg1, TArg2, TResult>(self: any, fn: (arg1: TArg1, arg2: TArg2, callback: Callback<TResult>) => void | any, arg1: TArg1, arg2: TArg2): Promise<[any, TResult | undefined]>;
export function promise<TResult>(self: any, fn: (...args: any[]) => any, ...args: any[]): Promise<[any, TResult | undefined]> {
    return new Promise((resolve, reject) => {
        fn.call(self, ...args, (err: any, result: TResult) => {
            resolve([err, result]);
        });
    });
}

export function call<TResult>(self: any, fn: () => TResult): [any, TResult | undefined];
export function call<TArg1, TResult>(self: any, fn: (arg1: TArg1) => TResult, arg1: TArg1): [any, TResult | undefined];
export function call<TArg1, TArg2, TResult>(self: any, fn: (arg1: TArg1, arg2: TArg2) => TResult, arg1: TArg1, arg2: TArg2): [any, TResult | undefined];
export function call<TResult>(self: any, fn: (...args: any[]) => any, ...args: any[]): [any, TResult | undefined] {
    try {
        const result = fn.call(self, ...args);
        return [undefined, result];
    } catch (err) {
        return [err, undefined];
    }
}

export function autobind<T extends object>(self: T): T {
    for (const key in self) {
        const value = (<any>self)[key];
        if (typeof value === 'function') {
            (<any>self)[key] = value.bind(self);
        }
    }
    return self;
}