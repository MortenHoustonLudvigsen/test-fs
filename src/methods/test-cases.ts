import { TestFs, Entry } from 'test-fs';
import { createFs, autobind } from '../test';
import { invariant } from 'test-fs/utils';

export abstract class TestCaseBase {
    readonly event: string;
    readonly eventCount: number = 1;

    constructor(
        readonly expected: any,
        readonly name: string,
        pathOrFd: string | number,
        readonly _init: (fs: TestFs, path: string) => void = (() => { })
    ) {
        if (typeof pathOrFd === 'string') {
            this.path = pathOrFd;
        }
        if (typeof pathOrFd === 'number') {
            this.fd = pathOrFd;
        }
        autobind(this);
    }

    fs: TestFs;
    entries: Entry[] = [];
    readonly path: string;
    readonly fd: number;

    init(): TestFs {
        this.entries = [];
        this.fs = createFs(true, 8, this.event);
        if (this.path) {
            this._init(this.fs, this.path);
        }
        if (this.event) {
            this.fs.on(this.event, entry => this.entries.push(entry));
        }
        return this.fs;
    }

    getResult(result: any): any {
        return result;
    }

    verify(err?: any, result?: any) {
        if (this.expected instanceof RegExp) {
            expect(result).toBeUndefined();
            expect(err).not.toBeUndefined();
            expect(err).toBeInstanceOf(Error);
            expect(err.message).toMatch(this.expected);
        } else {
            expect(err).toBeUndefined();
            this.verifyResult(this.getResult(result));
        }
    }

    abstract verifyResult(result: any): void;

    verifyEvents() {
        if (this.event) {
            if (this.expected instanceof RegExp) {
                expect(this.entries.length).toBe(0);
            } else {
                expect(this.entries.length).toBe(this.eventCount);
                expect(this.entries).toMatchSnapshot();
            }
        }
    }

    get expectedString(): string {
        return `return ${this.expected} for`;
    }

    get title(): string {
        if (this.expected instanceof RegExp) {
            return `should throw an ${this.expected.source} error for ${this.name}`;
        } else {
            return `should ${this.expectedString} ${this.name}`;
        }
    }

    get eventsTitle(): string {
        if (this.expected instanceof RegExp) {
            return `should not emit any '${this.event}' events for ${this.name}`;
        } else {
            const eventStr = this.eventCount === 1 ? `one '${this.event}' event` : `${this.eventCount} '${this.event}' events`;
            return `should emit ${eventStr} for ${this.name}`;
        }
    }
}

export class SnapshotTestCase extends TestCaseBase {
    constructor(
        readonly expected: 'snapshot' | RegExp,
        name: string,
        pathOrFd: string | number,
        _init: (fs: TestFs, path: string) => void = (() => { })
    ) {
        super(expected, name, pathOrFd, _init);
    }

    verifyResult(results: any) {
        expect(results).not.toBeUndefined();
        expect(results).toMatchSnapshot();
    }
}

export class ValueTestCase<TExpected> extends TestCaseBase {
    constructor(
        readonly expected: TExpected | RegExp,
        name: string,
        pathOrFd: string | number,
        _init: (fs: TestFs, path: string) => void = (() => { })
    ) {
        super(expected, name, pathOrFd, _init);
    }

    get expectedString(): string {
        return `return ${this.expected} for`;
    }

    verifyResult(results: any) {
        expect(results).toEqual(this.expected);
    }
}

export class NotSupportedTestCase extends TestCaseBase {
    readonly event = '';

    constructor(
        readonly methodName: string,
        readonly isAsync: boolean,
        pathOrFd: string | number,
        ...args: any[]
    ) {
        super(/ENOTSUP/, 'any entry', pathOrFd);
        this.method = (<any>TestFs.prototype)[methodName];
        this.args = [pathOrFd, ...args];
    }

    readonly method: Function;
    readonly args: any[];

    init(): TestFs {
        invariant(typeof this.method === 'function', `method must be a function`, this.methodName);
        return super.init();
    }

    verifyResult(results: any) {
        fail('Must throw ENOTSUP error');
    }

    verify(err?: any, result?: any) {
        expect(result).toBeUndefined();
        expect(err).not.toBeUndefined();
        expect(err).toBeInstanceOf(Error);
        expect(err.message).toMatch(this.expected);
    }
}
