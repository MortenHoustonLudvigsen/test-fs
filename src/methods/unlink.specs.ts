import { TestFs } from 'test-fs';
import { testPaths, call, promise } from '../test';
import { SnapshotTestCase } from './test-cases';

class TestCase extends SnapshotTestCase {
    readonly event = 'unlink';

    constructor(
        readonly expected: 'snapshot' | RegExp,
        readonly eventCount: number,
        name: string,
        path: string,
        _init: (fs: TestFs, path: string) => void = (() => { })
    ) {
        super(expected, name, path, _init);
    }
    
    get expectedString(): string {
        return 'delete';
    }

    getResult(result: any): any {
        return this.fs.getEntry(this.path);
    }

    verifyResult(results: any) {
        expect(results).not.toBeUndefined();
        expect(results.type).toBe('none');
        expect(results).toMatchSnapshot();
    }
}

const testCases = [
    new TestCase(/ENOENT/, 0, 'a directory, in a directory, that does not exist', testPaths.virtualFileInVirtualDir),
    new TestCase(/ENOENT/, 0, 'a directory, in a file', testPaths.virtualFileInPhysicalFile),
    new TestCase(/ENOENT/, 0, 'a directory, that does not exist', testPaths.virtualDir),
    new TestCase('snapshot', 1, 'a physical file', testPaths.physicalFile),
    new TestCase('snapshot', 1, 'a virtual file', testPaths.virtualFile, (fs, path) => fs.writeFileSync(path, 'dummy')),
    new TestCase('snapshot', 5, 'a directory, that is not empty', testPaths.physicalDir),
    new TestCase('snapshot', 1, 'a physical directory, that is empty', testPaths.emptyPhysicalDir),
    new TestCase('snapshot', 1, 'a virtual directory, that is empty', testPaths.virtualDir, (fs, path) => fs.mkdirSync(path)),
];

testCases.forEach(testCase => {
    it(`unlinkSync() ${testCase.title}`, () => {
        const fs = testCase.init();
        const [err, result] = call(fs, fs.unlinkSync, testCase.path);
        testCase.verify(err, result);
    });
});

testCases.forEach(testCase => {
    it(`unlinkSync() ${testCase.eventsTitle}`, () => {
        const fs = testCase.init();
        call(fs, fs.unlinkSync, testCase.path);
        testCase.verifyEvents();
    });
});

testCases.forEach(testCase => {
    it(`unlink() ${testCase.title}`, async () => {
        const fs = testCase.init();
        const [err, result] = await promise(fs, fs.unlink, testCase.path);
        testCase.verify(err, result);
    });
});

testCases.forEach(testCase => {
    it(`unlink() ${testCase.eventsTitle}`, async () => {
        const fs = testCase.init();
        await promise(fs, fs.unlink, testCase.path);
        testCase.verifyEvents();
    });
});
