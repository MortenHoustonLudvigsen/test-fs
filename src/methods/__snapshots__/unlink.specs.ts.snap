// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unlink() should delete a directory, that is not empty 1`] = `
<test-files>
    name.....: 'test-files'
    dir......: '.'
    type.....: 'none'
    version..: 2
`;

exports[`unlink() should delete a physical directory, that is empty 1`] = `
<test-files/emptydir>
    name.....: 'emptydir'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2
`;

exports[`unlink() should delete a physical file 1`] = `
<test-files/file.txt>
    name.....: 'file.txt'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2
`;

exports[`unlink() should delete a virtual directory, that is empty 1`] = `
<test-files/dummyDir>
    name.....: 'dummyDir'
    dir......: 'test-files'
    type.....: 'none'
    version..: 3
`;

exports[`unlink() should delete a virtual file 1`] = `
<test-files/dummy.txt>
    name.....: 'dummy.txt'
    dir......: 'test-files'
    type.....: 'none'
    version..: 3
`;

exports[`unlink() should emit 5 'unlink' events for a directory, that is not empty 1`] = `
Array [
  <test-files/emptydir>
    name.....: 'emptydir'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2,
  <test-files/file.bin>
    name.....: 'file.bin'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2,
  <test-files/file.txt>
    name.....: 'file.txt'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2,
  <test-files/file2.txt>
    name.....: 'file2.txt'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2,
  <test-files>
    name.....: 'test-files'
    dir......: '.'
    type.....: 'none'
    version..: 2,
]
`;

exports[`unlink() should emit one 'unlink' event for a physical directory, that is empty 1`] = `
Array [
  <test-files/emptydir>
    name.....: 'emptydir'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2,
]
`;

exports[`unlink() should emit one 'unlink' event for a physical file 1`] = `
Array [
  <test-files/file.txt>
    name.....: 'file.txt'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2,
]
`;

exports[`unlink() should emit one 'unlink' event for a virtual directory, that is empty 1`] = `
Array [
  <test-files/dummyDir>
    name.....: 'dummyDir'
    dir......: 'test-files'
    type.....: 'none'
    version..: 3,
]
`;

exports[`unlink() should emit one 'unlink' event for a virtual file 1`] = `
Array [
  <test-files/dummy.txt>
    name.....: 'dummy.txt'
    dir......: 'test-files'
    type.....: 'none'
    version..: 3,
]
`;

exports[`unlinkSync() should delete a directory, that is not empty 1`] = `
<test-files>
    name.....: 'test-files'
    dir......: '.'
    type.....: 'none'
    version..: 2
`;

exports[`unlinkSync() should delete a physical directory, that is empty 1`] = `
<test-files/emptydir>
    name.....: 'emptydir'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2
`;

exports[`unlinkSync() should delete a physical file 1`] = `
<test-files/file.txt>
    name.....: 'file.txt'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2
`;

exports[`unlinkSync() should delete a virtual directory, that is empty 1`] = `
<test-files/dummyDir>
    name.....: 'dummyDir'
    dir......: 'test-files'
    type.....: 'none'
    version..: 3
`;

exports[`unlinkSync() should delete a virtual file 1`] = `
<test-files/dummy.txt>
    name.....: 'dummy.txt'
    dir......: 'test-files'
    type.....: 'none'
    version..: 3
`;

exports[`unlinkSync() should emit 5 'unlink' events for a directory, that is not empty 1`] = `
Array [
  <test-files/emptydir>
    name.....: 'emptydir'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2,
  <test-files/file.bin>
    name.....: 'file.bin'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2,
  <test-files/file.txt>
    name.....: 'file.txt'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2,
  <test-files/file2.txt>
    name.....: 'file2.txt'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2,
  <test-files>
    name.....: 'test-files'
    dir......: '.'
    type.....: 'none'
    version..: 2,
]
`;

exports[`unlinkSync() should emit one 'unlink' event for a physical directory, that is empty 1`] = `
Array [
  <test-files/emptydir>
    name.....: 'emptydir'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2,
]
`;

exports[`unlinkSync() should emit one 'unlink' event for a physical file 1`] = `
Array [
  <test-files/file.txt>
    name.....: 'file.txt'
    dir......: 'test-files'
    type.....: 'none'
    version..: 2,
]
`;

exports[`unlinkSync() should emit one 'unlink' event for a virtual directory, that is empty 1`] = `
Array [
  <test-files/dummyDir>
    name.....: 'dummyDir'
    dir......: 'test-files'
    type.....: 'none'
    version..: 3,
]
`;

exports[`unlinkSync() should emit one 'unlink' event for a virtual file 1`] = `
Array [
  <test-files/dummy.txt>
    name.....: 'dummy.txt'
    dir......: 'test-files'
    type.....: 'none'
    version..: 3,
]
`;
